analyzers/src/SonarAnalyzer.Common/Rules/CollectionEmptinessCheckingBase.cs

```diff
@@ -0,0 +1,91 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class CollectionEmptinessCheckingBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+        where TSyntaxKind : struct
+    {
+        internal const string DiagnosticId = "S1155";
+
+        protected CollectionEmptinessCheckingBase() : base(DiagnosticId) { }
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer,
+                c =>
+                {
+                    var binaryLeft = Language.Syntax.BinaryExpressionLeft(c.Node);
+                    var binaryRight = Language.Syntax.BinaryExpressionRight(c.Node);
+
+                    if (Language.ExpressionNumericConverter.TryGetConstantIntValue(binaryLeft, out var left))
+                    {
+                        CheckExpression(c, binaryRight, left, Language.Syntax.ComparisonKind(c.Node).Mirror());
+                    }
+                    else if (Language.ExpressionNumericConverter.TryGetConstantIntValue(binaryRight, out var right))
+                    {
+                        CheckExpression(c, binaryLeft, right, Language.Syntax.ComparisonKind(c.Node));
+                    }
+                },
+                Language.SyntaxKind.ComparisonKinds);
+
+        private void CheckExpression(SyntaxNodeAnalysisContext context, SyntaxNode expression, int constant, ComparisonKind comparison)
+        {
+            if (comparison.Compare(constant).IsEmptyOrNotEmpty()
+                && TryGetCountCall(expression, context.SemanticModel, out var location, out var typeArgument))
+            {
+                context.ReportIssue(Diagnostic.Create(Rule, location, typeArgument));
+            }
+        }
+
+        private bool TryGetCountCall(SyntaxNode expression, SemanticModel semanticModel, out Location countLocation, out string typeArgument)
+        {
+            countLocation = null;
+            typeArgument = null;
+
+            if (Language.Syntax.NodeIdentifier(expression) is { } identifier
+                && identifier.ValueText == nameof(Enumerable.Count)
+                && (semanticModel.GetSymbolInfo(identifier.Parent.Parent).Symbol is IMethodSymbol methodSymbol)
+                && IsMethodCountExtension(methodSymbol)
+                && methodSymbol.IsExtensionOn(KnownType.System_Collections_Generic_IEnumerable_T))
+            {
+                if (methodSymbol.IsGenericMethod)
+                {
+                    typeArgument = methodSymbol.TypeArguments.Single().ToDisplayString();
+                }
```

Comment:
This should be something like this (not tested):
```
typeArgument = methodSymbol.ReceiverType.TypeArguments.Single().ToDisplayString();
```
Because
* we are sure we're on `System_Collections_Generic_IEnumerable_T`
* C# method is generic (always true anyway), but VB is not 

```
System.Collections.Generic.IEnumerable<string>.Count<string>
System.Collections.Generic.IEnumerable(Of System.String).Count()
```

Reply:
@pavel-mikula-sonarsource 
It turns out that OR ReceiverType OR methodSymbol contains the generic part. I changed it. Both work fine now. :)